File: Lab08/CO523_Lab08_E20266/concept_implementation/C_pp/polymorphism_method_overloading.cpp
// CO523 - Programming Languages | Lab 08: Object-Oriented Programming Paradigm
// Section 3.4.1 - Compile-time Polymorphism (Method Overloading)
//
// Method overloading allows a class to have multiple methods with the same name
// but different parameter lists. The correct method is chosen at compile time
// based on the number and type of arguments passed.

#include <iostream>
#include <string>

class Employee {
public:
    std::string name;
    int id;
    double salary;

    // Overload 1: work() with no arguments — general work message
    void work() const {
        std::cout << name << " is working..." << '\n';
    }

    // Overload 2: work(project) — specific work message with a project name
    void work(const std::string& project) const {
        std::cout << name << " is working on project: " << project << '\n';
    }
};

int main() {
    Employee emp;
    emp.name = "Jeorge";
    emp.id = 1452;
    emp.salary = 71000;

    emp.work();                    // Calls overload 1 — no argument
    emp.work("Website Redesign");  // Calls overload 2 — string argument

    return 0;
}

File: Lab08/CO523_Lab08_E20449/concept_implementation/python/polymorphism_method_overloading.py
# CO523 - Programming Languages | Lab 08: Object-Oriented Programming Paradigm
# Section 3.4.1 - Compile-time Polymorphism (Method Overloading)
#
# Python does not natively support method overloading, but the same effect is achieved
# using a default parameter (project=None). The method behaves differently based on
# whether a project name is passed or not.

class Employee:
    def __init__(self):
        self.name = ""
        self.id = 0
        self.salary = 0.0

    # Single method simulating overloading via an optional parameter
    def work(self, project=None):
        if project is None:
            # No argument — general work message (simulates overload 1)
            print(f"{self.name} is working...")
        else:
            # Project argument given — specific work message (simulates overload 2)
            print(f"{self.name} is working on project: {project}")


def main():
    emp = Employee()
    emp.name = "Jeorge"
    emp.id = 1452
    emp.salary = 71000

    emp.work()                    # Called without argument
    emp.work("Website Redesign")  # Called with a project argument


if __name__ == "__main__":
    main()
